File: ios/GuideStreamTV/Views/HouseSlotCard.swift
//
//  HouseSlotCard.swift
//  GuideStreamTV
//
//  Last-resort fill for a sponsored slot: a house card promoting the app's
//  own features.
//
//  WHY THIS EXISTS. A slot picks an affiliate offer from a pool of eight
//  services, filtered to the ones the viewer does NOT already subscribe to —
//  the app must never advertise something they already pay for. A viewer who
//  subscribes to all eight therefore had no eligible offer anywhere, and the
//  slot fell through to `allowRakutenFallback == false`, whose only remaining
//  content was an AdMob native ad. On an AdMob miss the slot collapsed to zero
//  height, so the most engaged users saw nothing in any slot in the entire app
//  and it read as "advertising is broken".
//
//  A house card keeps the promise (never advertise an owned service) without
//  the dead space. It is the LAST resort: an eligible affiliate offer wins,
//  and a native ad still upgrades the slot the moment one lands in the pool.
//

import SwiftUI

struct HouseSlotCard: View {
    struct Offer {
        let icon: String
        let headline: String
        let subtitle: String
    }

    let offer: Offer
    var compact: Bool = false
    var feedStyle: Bool = false
    var onDismiss: () -> Void

    /// House offers describe things the app already does, so none of them can
    /// promise something a viewer cannot find.
    private static let pool: [Offer] = [
        Offer(icon: "square.stack.3d.up.fill",
              headline: "One watchlist, every service",
              subtitle: "Save a show once and see it wherever you watch"),
        Offer(icon: "bell.badge.fill",
              headline: "Never miss a premiere",
              subtitle: "Get told the day a new episode lands"),
        Offer(icon: "sportscourt.fill",
              headline: "Follow your teams",
              subtitle: "Live scores and where the game is on"),
        Offer(icon: "sparkles.tv.fill",
              headline: "Watch on the big screen",
              subtitle: "GuideStream TV is on Apple TV too")
    ]

    static func offer(for seed: Int) -> Offer {
        pool[abs(seed) % pool.count]
    }

    var body: some View {
        HStack(spacing: 12) {
            RoundedRectangle(cornerRadius: 10, style: .continuous)
                .fill(Color.orange.opacity(0.16))
                .frame(width: iconPlate, height: iconPlate)
                .overlay(
                    Image(systemName: offer.icon)
                        .scaledFont(size: iconPlate * 0.46, weight: .semibold)
                        .foregroundStyle(Color.orange)
                )

            VStack(alignment: .leading, spacing: 2) {
                Text(offer.headline)
                    .scaledFont(size: compact ? 13 : 14, weight: .semibold)
                    .foregroundStyle(Color.textPrimary)
                    .lineLimit(1)
                Text(offer.subtitle)
                    .scaledFont(size: compact ? 11 : 12, weight: .regular)
                    .foregroundStyle(Color.textSecondary)
                    .lineLimit(2)
            }

            Spacer(minLength: 8)

            Button(action: onDismiss) {
                Image(systemName: "xmark")
                    .scaledFont(size: 10, weight: .bold)
                    .foregroundStyle(Color.textTertiary)
                    .frame(width: 28, height: 28)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Dismiss")
        }
        .padding(.horizontal, 12)
        .frame(height: cardHeight)
        .frame(maxWidth: .infinity)
        .background(
            RoundedRectangle(cornerRadius: 14, style: .continuous)
                .fill(Color.white.opacity(0.06))
        )
        .overlay(
            RoundedRectangle(cornerRadius: 14, style: .continuous)
                .stroke(Color.white.opacity(0.08), lineWidth: 1)
        )
        .accessibilityElement(children: .combine)
        .accessibilityLabel("\(offer.headline). \(offer.subtitle)")
    }

    /// Matches the 96pt inline chip row and the taller detail-sheet card, so
    /// swapping in a native ad causes no layout jump.
    private var cardHeight: CGFloat { compact && feedStyle ? 96 : 84 }
    private var iconPlate: CGFloat { compact && feedStyle ? 56 : 44 }
}

#Preview("House slot") {
    ZStack {
        Color.navy.ignoresSafeArea()
        VStack(spacing: 16) {
            HouseSlotCard(offer: HouseSlotCard.offer(for: 0), compact: true, feedStyle: true, onDismiss: {})
            HouseSlotCard(offer: HouseSlotCard.offer(for: 1), onDismiss: {})
        }
        .padding()
    }
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/ui/ads/HouseSlotCard.kt
package com.rork.guidestreamtvandroid.ui.ads

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.filled.Tv
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.rork.guidestreamtvandroid.ui.theme.BrandOrange
import com.rork.guidestreamtvandroid.ui.theme.TextPrimary
import com.rork.guidestreamtvandroid.ui.theme.TextSecondary

/**
 * Last-resort fill for a sponsored slot: a house card promoting the app's own
 * features. Mirrors iOS HouseSlotCard.swift.
 *
 * WHY THIS EXISTS. A slot picks an affiliate offer from a pool of eight
 * services, filtered to the ones the viewer does NOT already subscribe to —
 * the app must never advertise something they already pay for. A viewer who
 * subscribes to all eight therefore had no eligible offer anywhere, and
 * `if (!allowRakutenFallback && adMobFailed) return` rendered nothing at all,
 * so the most engaged users saw empty slots across the whole app.
 *
 * This keeps the promise without the dead space. It is the LAST resort: an
 * eligible affiliate offer wins, and a native ad still takes the slot when one
 * fills.
 */
data class HouseOffer(
    val icon: ImageVector,
    val headline: String,
    val subtitle: String,
)

private val HOUSE_POOL = listOf(
    HouseOffer(Icons.Filled.PlayArrow, "One watchlist, every service",
        "Save a show once and see it wherever you watch"),
    HouseOffer(Icons.Filled.Notifications, "Never miss a premiere",
        "Get told the day a new episode lands"),
    HouseOffer(Icons.Filled.Star, "Follow your teams",
        "Live scores and where the game is on"),
    HouseOffer(Icons.Filled.Tv, "Watch on the big screen",
        "GuideStream TV is on Android TV too"),
)

fun houseOfferFor(seed: String): HouseOffer =
    HOUSE_POOL[(kotlin.math.abs(seed.hashCode())) % HOUSE_POOL.size]

@Composable
fun HouseSlotCard(
    offer: HouseOffer,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
    feedStyle: Boolean = true,
) {
    val cardHeight = if (feedStyle) 96.dp else 84.dp
    val plate = if (feedStyle) 56.dp else 44.dp

    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(cardHeight)
            .clip(RoundedCornerShape(14.dp))
            .background(Color.White.copy(alpha = 0.06f))
            .border(1.dp, Color.White.copy(alpha = 0.08f), RoundedCornerShape(14.dp))
            .padding(horizontal = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Box(
            modifier = Modifier
                .size(plate)
                .clip(RoundedCornerShape(10.dp))
                .background(BrandOrange.copy(alpha = 0.16f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = offer.icon,
                contentDescription = null,
                tint = BrandOrange,
                modifier = Modifier.size(plate * 0.46f),
            )
        }

        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = offer.headline,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                color = TextPrimary,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                text = offer.subtitle,
                fontSize = 12.sp,
                color = TextSecondary,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
        }

        Box(
            modifier = Modifier
                .size(28.dp)
                .clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null,
                ) { onDismiss() },
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = Icons.Filled.Close,
                contentDescription = "Dismiss",
                tint = Color.White.copy(alpha = 0.35f),
                modifier = Modifier.size(14.dp),
            )
        }
    }
}
